### PR TITLE
MDL-73763 core : Index fixes for question_references and question_set…

### PR DESCRIPTION
**Pull request by marcghaly**
Hi @safatshahin,
I have made a new branch with clean fixes - tested the upgrade.

We should discuss this comment if possible (from tracker) `Arguably, there should also be a foreign key of (questionbankentryid, version) pointing at question_versions.` 

Cheers !

### CODE REVIEW

File: lib/db/upgrade.php [hunk 1]
```diff
@@ -3881,6 +3881,14 @@ function xmldb_main_upgrade($oldversion) {
             $dbman->create_table($table);
         }
 
+        $index = new xmldb_index('usingcontextid-component-questionarea-itemid', XMLDB_INDEX_UNIQUE,
```
Comment (safatshahin) [inline lib/db/upgrade.php:3884]: this is not the correct location for this change, as this part is already released, create a new upgrade logic for this.
Also the name is not good, name the variable something relevant to the table.

File: lib/db/upgrade.php [hunk 2]
```diff
@@ -3903,6 +3911,14 @@ function xmldb_main_upgrade($oldversion) {
             $dbman->create_table($table);
         }
 
+        $index = new xmldb_index('usingcontextid, component, questionarea, itemid', XMLDB_INDEX_UNIQUE,
```
Comment (safatshahin) [inline lib/db/upgrade.php:3914]: this is not the correct location for this change, as this part is already released, create a new upgrade logic for this.
Also the name is not good, name the variable something relevant to the table.

File: lib/db/upgrade.php (hunk 1 above)
Comment (marcghaly) [reply lib/db/upgrade.php]: Changed variable name and added a new logic.

File: lib/db/upgrade.php (hunk 2 above)
Comment (marcghaly) [reply lib/db/upgrade.php]: Changed variable name and added a new logic.

File: lib/db/upgrade.php [hunk 3]
```diff
@@ -3907,6 +3907,40 @@ function xmldb_main_upgrade($oldversion) {
         upgrade_main_savepoint(true, 2022020200.01);
     }
 
+    // Unique indexes for question_set_references.
+    if ($oldversion < 2022020500.00) {
```
Comment (safatshahin) [inline lib/db/upgrade.php:3911]: it can be done in a same chunk, please merge them to one.

File: lib/db/upgrade.php [hunk 4]
```diff
@@ -3907,6 +3907,40 @@ function xmldb_main_upgrade($oldversion) {
         upgrade_main_savepoint(true, 2022020200.01);
     }
 
+    // Unique indexes for question_set_references.
```
Comment (safatshahin) [inline lib/db/upgrade.php:3910]: add this to the end of the file, following the pattern like the previous ones, like a unique savepoint and oldversion which greater than the current version of core

File: lib/db/upgrade.php (hunk 3 above)
Comment (marcghaly) [reply lib/db/upgrade.php]: Done.

File: lib/db/upgrade.php (hunk 4 above)
Comment (marcghaly) [reply lib/db/upgrade.php]: Added in date order.

**Comment by marcghaly** [comment]
Hi @safatshahin,
I have made changes as requested,
Cheers

**Comment by safatshahin** [comment]
Tracker for this issue:
https://tracker.moodle.org/browse/MDL-73763

**Comment by tuanngocnguyen** [comment]
I have rebased to fix conflicts.

The MR looks good to me

**Comment by safatshahin** [comment]
Thanks for your review @tuanngocnguyen 
moving the review to tracker now.

**Pull request closed without merging**